library/src/scala/IArray.scala

```diff
@@ -249,9 +250,18 @@ object IArray:
   extension [T](arr: IArray[T]) def takeWhile(p: T => Boolean): IArray[T] =
     genericArrayOps(arr).takeWhile(p)
 
-  /** Returns a mutable copy of this immutable array. */
-  extension [T](arr: IArray[T]) def toArray: Array[T] =
-    arr.clone.asInstanceOf[Array[T]]
+  // NOTE: these two overloads are never selected when calling toArray on an
+  // IArray because they're considered ambiguous, but such calls still typecheck
+  // because of the implicit conversion `genericWrapArray`.
+  extension [T](arr: IArray[T])
+    /** Returns a mutable copy of this immutable array. */
+    @deprecated("This method implementation is incorrect and calling it can crash your program, please use the other overload.", "3.0.1")
+    def toArray: Array[T] =
+      arr.clone.asInstanceOf[Array[T]]
+    /** Returns a mutable copy of this immutable array. */
+    @experimental
```

Comment:
`@experimental` seems to become an excuse not to take patch versions seriously. It's fine to use it in some obscure corner of the language, where it's unlikely to show up in anyone's autocompletion or documentation unless they're looking for it. It's not fine to use it on common collection types with common (and existing!) names that will pop up in everyone's autocomplete list.

Also the fact that it won't actually be chosen does not excuse the `@experimental`. Its very *presence* causes something else not to be selected anymore. So it's even worse: now you've got an `@experimental` feature that causes non experimental code to compile in a different way, *silently*.

This change simply needs to wait for 3.1.0, instead of being precipitated in 3.0.1 under the cover of `@experimental`. (Or it might be the reason to publish a 3.1.0 sooner than expected.)

Reply:
I'm not sure I understand what you're proposing exactly. At the very least we should clearly mark the broken method as broken, so it needs to be deprecated. Then we could not add the experimental overload and wait for 3.1.0 to add it as experimental, but I don't know what that would gain us, in 3.1.0 it still will be ambiguous with the deprecated method and we'll end up selecting the one from the implicit conversion, given that we can also do this right now without breaking either source or binary compatibility, I don't see why we shouldn't.

Reply:
(Ideally, I think we should have some mechanism like the old trick of making a method package-private to hide it but that would preserve both tasty- and binary-compatibility, then we could hide this method and still get source-compatibility from the implicit conversion).

Reply:
* It's not selected *if* you rely on the extension; it can still be called explicitly
* It breaks forward binary compatibility if it's not experimental (and it shouldn't be)
* It breaks source compatibility because now the method that is selected by default requires a `ClassTag`, which some call sites may not have

This can't be fixed in a patch version. This is the price to pay for any sort of compatibility. Some things can't be fixed in patch versions. Some things can't even be fixed in minor versions.

If this needs to be fixed ASAP, then release a minor version ASAP. There's not much downside to it. Except having to tell to users "hey maybe this breaks source compat in rare situations". Which if you release the current fix in a patch versions, you're still breaking source compat in rare situations but you're not signaling that to your users with a minor version bump.

Reply:
So to be clear: I'm proposing not to do anything about this in a patch version, and instead release the current proposed fix, without `@experimental`, in a minor version.

Reply:
We've never gone through the process of making a minor version yet and I think we'll have to take some time to discuss it first (e.g. do we need to bump the tasty version too? if we don't, what's stopping a scala 3.0 project from depending on a 3.1 macro and crashing at runtime because the macro calls a 3.1 API that it won't find on the 3.0 standard library on the JVM classpath?). At the same time we shouldn't let people call a broken method without warnings, so if you're opposed to adding the overload, I'll just deprecate the method with a message providing an alternative.

Reply:
> At the same time we shouldn't let people call a broken method without warnings, so if you're opposed to adding the overload, I'll just deprecate the method with a message providing an alternative.

OK, let's do that for now.

Reply:
Alright, done.